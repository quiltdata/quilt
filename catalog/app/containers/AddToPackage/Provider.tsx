import { basename } from 'path'

import * as R from 'ramda'
import * as React from 'react'

import type * as Model from 'model'

const Ctx = React.createContext<{
  append: (file: Model.S3File) => void
  clear: () => void
  entries: Record<string, Model.S3File>
  merge: (entries: Record<string, Model.S3File>) => void
} | null>(null)

interface ProviderProps {
  children: React.ReactNode
}

/** @deprecated Use Bookmarks.Provider and 'addToPackage' GroupName */
export function Provider({ children }: ProviderProps) {
  const [entries, setEntries] = React.useState({})
  const append = React.useCallback((s3File) => {
    setEntries(
      R.mergeLeft({
        [basename(s3File.key)]: s3File,
      }),
    )
  }, [])
  const clear = React.useCallback(() => setEntries({}), [])
  const merge = React.useCallback((files) => setEntries(R.mergeLeft(files)), [])

  const value = React.useMemo(
    () => ({ append, clear, entries, merge }),
    [append, clear, entries, merge],
  )

  return <Ctx.Provider value={value}>{children}</Ctx.Provider>
}

/** @deprecated Use useBookmarks and 'addToPackage' GroupName */
const useAddToPackage = () => React.useContext(Ctx)

/** @deprecated Use Bookmarks.use and 'addToPackage' GroupName */
export const use = useAddToPackage
