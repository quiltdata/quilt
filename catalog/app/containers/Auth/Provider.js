import { fromJS } from 'immutable';
import pick from 'lodash/fp/pick';
import PT from 'prop-types';
import { Fragment } from 'react';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import { mapProps, withHandlers, withProps, setPropTypes } from 'recompose';

import { push as notify } from 'containers/Notifications/actions';
import { composeComponent } from 'utils/reactTools';
import { injectReducer } from 'utils/ReducerInjector';
import { injectSaga } from 'utils/SagaInjector';
import * as storage from 'utils/storage';

import { REDUX_KEY } from './constants';
import msg from './messages';
import reducer from './reducer';
import saga from './saga';


const ActionPattern = PT.oneOfType([PT.string, PT.func]);

export default composeComponent('Auth.Provider',
  injectIntl,
  connect(undefined, undefined, undefined, { pure: false }),
  setPropTypes({
    intl: PT.shape({
      formatMessage: PT.func.isRequired,
    }).isRequired,
    dispatch: PT.func.isRequired,
    checkOn: PT.oneOfType([ActionPattern, PT.arrayOf(ActionPattern)]),
  }),
  withProps({
    storeTokens: (tokens) => storage.set('tokens', tokens),
    forgetTokens: () => storage.remove('tokens'),
    storeUser: (user) => storage.set('user', user),
    forgetUser: () => storage.remove('user'),
  }),
  withHandlers({
    onAuthLost: ({ intl, dispatch }) => () => {
      dispatch(notify(intl.formatMessage(msg.notificationAuthLost)));
    },
    onAuthError: ({ intl, dispatch }) => () => {
      dispatch(notify(intl.formatMessage(msg.notificationAuthError)));
    },
  }),
  injectReducer(REDUX_KEY, reducer, () =>
    fromJS(storage.load()).filter(Boolean).update((s) =>
      s.set('state', s.getIn(['user', 'current_user']) ? 'SIGNED_IN' : 'SIGNED_OUT'))),
  injectSaga(REDUX_KEY, saga),
  mapProps(pick(['children'])),
  Fragment);
