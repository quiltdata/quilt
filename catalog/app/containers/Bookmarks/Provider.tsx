import * as R from 'ramda'
import * as React from 'react'

// TODO: rename all variables 'file' to 'handle'
import { S3HandleBase, handleToS3Url } from 'utils/s3paths'
import mkStorage from 'utils/storage'

const STORAGE_KEYS = {
  BOOKMARKS: 'BOOKMARKS',
}
const storage = mkStorage({
  [STORAGE_KEYS.BOOKMARKS]: STORAGE_KEYS.BOOKMARKS,
})

type GroupName = 'main'

interface BookmarksGroup {
  entries: Record<string, S3HandleBase>
}

type BookmarksGroups = Record<GroupName, BookmarksGroup>

const Ctx = React.createContext<{
  append: (groupName: GroupName, file: S3HandleBase | S3HandleBase[]) => void
  clear: (groupName: GroupName) => void
  groups: BookmarksGroups
  hasUpdates: boolean
  hide: () => void
  isBookmarked: (groupName: GroupName, file: S3HandleBase) => boolean
  isOpened: boolean
  remove: (groupName: GroupName, file: S3HandleBase) => void
  show: () => void
  toggle: (groupName: GroupName, file: S3HandleBase) => void
} | null>(null)

interface ProviderProps {
  children: React.ReactNode
}

const initialBookmarks = { main: { entries: {} } }

const keyResolver = (file: S3HandleBase) => handleToS3Url(file)

type StateUpdaterFunction = (input: BookmarksGroups) => BookmarksGroups

function createAppendUpdater(
  groupName: GroupName,
  s3File: S3HandleBase | S3HandleBase[],
): StateUpdaterFunction {
  if (Array.isArray(s3File)) {
    const entries = s3File.reduce(
      (memo, entry) => ({
        ...memo,
        [keyResolver(entry)]: entry,
      }),
      {},
    )
    return R.over(R.lensPath([groupName, 'entries']), R.mergeLeft(entries))
  }
  return R.set(R.lensPath([groupName, 'entries', keyResolver(s3File)]), s3File)
}

function createRemoveUpdater(
  groupName: GroupName,
  s3File: S3HandleBase,
): StateUpdaterFunction {
  return R.over(R.lensPath([groupName, 'entries']), R.dissoc(keyResolver(s3File)))
}

function createClearUpdater(groupName: GroupName): StateUpdaterFunction {
  return R.assocPath([groupName, 'entries'], {})
}

export function Provider({ children }: ProviderProps) {
  const [hasUpdates, setUpdates] = React.useState(false)
  const [isOpened, setOpened] = React.useState(false)
  const [groups, setGroups] = React.useState(() => {
    try {
      return storage.get(STORAGE_KEYS.BOOKMARKS) || initialBookmarks
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
      return initialBookmarks
    }
  })
  const updateGroups = React.useCallback(
    (updater: StateUpdaterFunction) => {
      const newGroups = updater(groups)
      setGroups(newGroups)
      storage.set(STORAGE_KEYS.BOOKMARKS, newGroups)
    },
    [groups],
  )
  const append = React.useCallback(
    (groupName: GroupName, s3File: S3HandleBase | S3HandleBase[]) => {
      updateGroups(createAppendUpdater(groupName, s3File))
      if (!isOpened) setUpdates(true)
    },
    [isOpened, updateGroups],
  )
  const remove = React.useCallback(
    (groupName: GroupName, s3File: S3HandleBase) => {
      const isLastBookmark =
        R.pipe(R.path([groupName, 'entries']), R.keys, R.length)(groups) === 1
      updateGroups(createRemoveUpdater(groupName, s3File))
      if (isLastBookmark) {
        setUpdates(false)
      } else if (!isOpened) {
        setUpdates(true)
      }
    },
    [groups, isOpened, updateGroups],
  )
  const clear = React.useCallback(
    (groupName: GroupName) => {
      updateGroups(createClearUpdater(groupName))
      setUpdates(false)
    },
    [updateGroups],
  )
  const isBookmarked = React.useCallback(
    (groupName: GroupName, s3File: S3HandleBase) =>
      R.hasPath([groupName, 'entries', keyResolver(s3File)], groups),
    [groups],
  )
  const toggle = React.useCallback(
    (groupName: GroupName, s3File: S3HandleBase) =>
      isBookmarked(groupName, s3File)
        ? remove(groupName, s3File)
        : append(groupName, s3File),
    [isBookmarked, remove, append],
  )
  const hide = React.useCallback(() => {
    setOpened(false)
    setUpdates(false)
  }, [])

  const value = React.useMemo(
    () => ({
      append,
      clear,
      groups,
      hasUpdates,
      hide,
      isBookmarked,
      isOpened,
      remove,
      show: () => setOpened(true),
      toggle,
    }),
    [append, clear, groups, hasUpdates, hide, isBookmarked, isOpened, remove, toggle],
  )

  return <Ctx.Provider value={value}>{children}</Ctx.Provider>
}

export const useBookmarks = () => React.useContext(Ctx)

export const use = useBookmarks
