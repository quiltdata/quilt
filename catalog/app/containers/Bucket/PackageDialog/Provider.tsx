import * as React from 'react'
import * as RRDom from 'react-router-dom'

import * as workflows from 'utils/workflows'

import {
  PackageCreationDialogUIOptions,
  PackageCreationDialog,
} from '../PackageDialog/PackageCreationForm'

export function CreatePackageLink(props: Omit<RRDom.LinkProps, 'to'>) {
  const location = RRDom.useLocation()
  const search = React.useMemo(() => {
    const searchParams = new URLSearchParams(location.search)
    searchParams.set('createPackage', 'true')
    return searchParams.toString()
  }, [location.search])
  return <RRDom.Link {...props} to={(loc) => ({ ...loc, search: `?${search}` })} />
}

const Ctx = React.createContext<{
  open: () => void
  setDst: (s: workflows.Successor) => void
} | null>(null)

interface ProviderProps {
  // Use it if you need to add CreatePackage dialog for specific places
  // ex. ?createPackage=bookmarks
  // and ignore global ?cretePackage=true
  id?: 'bookmarks'
  bucket: string
  name?: string
  hashOrTag?: string
  s3Path?: string
  ui: PackageCreationDialogUIOptions
  delayHashing?: boolean
  disableStateDisplay?: boolean
  children: React.ReactNode
}

export function Provider({
  id,
  bucket,
  name,
  hashOrTag,
  s3Path,
  ui,
  delayHashing,
  disableStateDisplay,
  children,
}: ProviderProps) {
  const history = RRDom.useHistory()
  const location = RRDom.useLocation()
  const searchParams = React.useMemo(
    () => new URLSearchParams(location.search),
    [location.search],
  )
  const open = React.useCallback(() => {
    searchParams.set('createPackage', id || 'true')
    history.push({
      search: searchParams.toString(),
    })
  }, [id, history, searchParams])
  const [successor, setSuccessor] = React.useState({
    slug: bucket,
  } as workflows.Successor)
  const packageHandle = React.useMemo(() => {
    if (!name || !hashOrTag) return undefined
    return { name, hashOrTag }
  }, [name, hashOrTag])
  const value = React.useMemo(
    () => ({
      open,
      setDst: setSuccessor,
    }),
    [open],
  )
  return (
    <Ctx.Provider value={value}>
      <PackageCreationDialog
        ui={ui}
        id={id}
        src={{ bucket, packageHandle, s3Path }}
        successor={successor}
        onSuccessor={setSuccessor}
        delayHashing={delayHashing}
        disableStateDisplay={disableStateDisplay}
      />
      {children}
    </Ctx.Provider>
  )
}

export const useCreatePackage = () => React.useContext(Ctx)

export const use = useCreatePackage
