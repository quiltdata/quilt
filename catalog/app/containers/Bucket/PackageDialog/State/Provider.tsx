import invariant from 'invariant'
import * as React from 'react'

import type { FilesState } from './files'
import type { PackageSrc } from './manifest'
import { State, useState } from './State'

interface ProviderProps {
  children: React.ReactNode
  src?: PackageSrc
  dst: { bucket: string; name?: string }
  open?: boolean | FilesState['value']['added']
}

export function Provider({ children, dst, src, open = false }: ProviderProps) {
  const contextValue = useState(dst, src, open)
  return <Context.Provider value={contextValue}>{children}</Context.Provider>
}

export const Context = React.createContext<State | null>(null)

export function useContext(): State {
  const context = React.useContext(Context)
  invariant(context, 'useContext must be used within PackageDialogProvider')
  return context
}
