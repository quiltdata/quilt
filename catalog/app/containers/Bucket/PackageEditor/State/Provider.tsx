import * as React from 'react'

import { L } from 'components/Form/Package/types'
import type * as Types from 'utils/types'
import type { Schema } from 'utils/workflows'

import type { Manifest } from '../../PackageDialog/Manifest'

import useSource, { Src } from './Source'
import useManifest from './Manifest'
import useBucket, { BucketContext } from './Bucket'
import useWorkflow, { WorkflowContext } from './Workflow'

export interface InputState {
  errors?: Error[] | typeof L
  value: string
}

export interface MetaState {
  value?: Types.JsonRecord
  schema?: Schema
}

interface MessageContext {
  state: InputState | typeof L
  actions: {
    onChange: (v: string) => void
  }
}

interface MetaContext {
  state: MetaState | typeof L
}

interface NameContext {
  state: InputState | typeof L
  actions: {
    onChange: (v: string) => void
  }
}

interface ContextData {
  bucket: BucketContext
  message: MessageContext
  meta: MetaContext
  name: NameContext
  workflow: WorkflowContext
}

function useMessage(): MessageContext {
  const [value, setValue] = React.useState('')
  return React.useMemo(
    () => ({
      state: {
        value,
      },
      actions: {
        onChange: setValue,
      },
    }),
    [value],
  )
}

function useName(src: Src, workflow: WorkflowContext): NameContext {
  const [value, setValue] = React.useState(src.packageHandle?.name || '')
  const state = React.useMemo(
    () => (workflow.state === L ? L : { value }),
    [value, workflow.state],
  )
  return React.useMemo(
    () => ({
      state,
      actions: {
        onChange: setValue,
      },
    }),
    [state],
  )
}

function useMeta(workflow: WorkflowContext, manifest?: Manifest | typeof L): MetaContext {
  const state = React.useMemo(() => {
    if (manifest === L || workflow.state === L) return L
    return {
      value: manifest?.meta,
      schema: workflow.state.value?.schema,
    }
  }, [workflow.state, manifest])
  return React.useMemo(
    () => ({
      state,
    }),
    [state],
  )
}

const Ctx = React.createContext<ContextData | null>(null)

export function useContext(): ContextData {
  const data = React.useContext(Ctx)
  if (!data) throw new Error('Set provider')
  return data
}

interface ProviderProps {
  bucket: string
  name: string
  hashOrTag: string
  hash?: string
  path: string
  mode?: string
  resolvedFrom?: string
  size?: number
  children: React.ReactNode
}

export default function Provider({
  bucket: srcBucket,
  name: srcName,
  hashOrTag,
  path,
  children,
}: ProviderProps) {
  const src = useSource(srcBucket, srcName, hashOrTag, path)
  const bucket = useBucket(src)

  const manifest = useManifest(src)
  const workflow = useWorkflow(bucket.state?.value, manifest)
  const name = useName(src, workflow)
  const message = useMessage()
  const meta = useMeta(workflow, manifest)
  const v = React.useMemo(
    () => ({
      bucket,
      message,
      name,
      workflow,
      meta,
    }),
    [bucket, message, meta, name, workflow],
  )
  return <Ctx.Provider value={v}>{children}</Ctx.Provider>
}
