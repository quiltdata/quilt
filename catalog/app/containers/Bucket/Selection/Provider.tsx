import * as R from 'ramda'
import * as React from 'react'

import { EMPTY_MAP, ListingSelection, SelectionItem, merge } from './utils'

interface State {
  inited: boolean
  clear: () => void
  isEmpty: boolean
  merge: (items: SelectionItem[], bucket: string, path: string, filter?: string) => void
  remove: (prefixUrl: string, index: number) => { isEmpty: boolean }
  selection: ListingSelection
  totalCount: number
}

const dummy = () => {
  new Error('Selection provider not initialized')
}

const Ctx = React.createContext<State>({
  inited: false,
  clear: dummy,
  isEmpty: true,
  merge: dummy,
  remove: () => {
    dummy()
    return { isEmpty: true }
  },
  selection: EMPTY_MAP,
  totalCount: 0,
})

interface ProviderProps {
  children: React.ReactNode
}

export function Provider({ children }: ProviderProps) {
  const [selection, setSelection] = React.useState<ListingSelection>(EMPTY_MAP)
  const totalCount = React.useMemo(
    () => Object.values(selection).reduce((acc, ids) => acc + ids.length, 0),
    [selection],
  )
  const handleMerge = React.useCallback(
    (ids, bucket, path, filter) => setSelection(merge(ids, bucket, path, filter)),
    [],
  )
  const clear = React.useCallback(() => setSelection(EMPTY_MAP), [])
  const remove = React.useCallback(
    (prefixUrl: string, index: number) => {
      const newSelection = R.dissocPath<ListingSelection>([prefixUrl, index], selection)
      setSelection(newSelection)
      return {
        isEmpty: !Object.values(newSelection).some((ids) => !!ids.length),
      }
    },
    [selection],
  )
  const state = React.useMemo(
    () => ({
      inited: true,
      clear,
      isEmpty: totalCount === 0,
      merge: handleMerge,
      remove,
      selection,
      totalCount,
    }),
    [clear, handleMerge, remove, selection, totalCount],
  )
  return <Ctx.Provider value={state}>{children}</Ctx.Provider>
}

export const useSelection = () => React.useContext(Ctx)

export const use = useSelection
