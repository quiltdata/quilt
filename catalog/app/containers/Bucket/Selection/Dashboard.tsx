import { basename } from 'path'

import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import * as Bookmarks from 'containers/Bookmarks/Provider'
import type * as Model from 'model'
import * as NamedRoutes from 'utils/NamedRoutes'
import * as PackageUri from 'utils/PackageUri'
import StyledLink from 'utils/StyledLink'
import type { PackageHandle } from 'utils/packageHandle'
import * as s3paths from 'utils/s3paths'

import * as FileView from '../FileView'

import { useSelection } from './Provider'
import { toHandlesMap, toHandlesList } from './utils'

const useEmptyStateStyles = M.makeStyles((t) => ({
  root: {
    padding: t.spacing(3, 0, 2),
  },
}))

function EmptyState() {
  const classes = useEmptyStateStyles()
  return (
    <M.Typography className={classes.root} variant="h5">
      Nothing selected
    </M.Typography>
  )
}

interface FileLinkProps {
  className: string
  handle: Model.S3.S3ObjectLocation
  packageHandle?: PackageHandle
}

function FileLink({ className, handle, packageHandle }: FileLinkProps) {
  const { urls } = NamedRoutes.use()

  if (!packageHandle) {
    const to = s3paths.isDir(handle.key)
      ? urls.bucketDir(handle.bucket, handle.key)
      : urls.bucketFile(handle.bucket, handle.key)
    const children = decodeURIComponent(s3paths.handleToS3Url(handle))
    return <StyledLink {...{ children, className, to }} />
  }

  const to = urls.bucketPackageTree(
    packageHandle.bucket,
    packageHandle.name,
    packageHandle.hash,
    handle.key,
  )
  const packageUri = PackageUri.stringify({
    ...packageHandle,
    path: handle.key,
    catalog: window.location.hostname,
  })
  const children = decodeURIComponent(
    packageUri.slice(0, packageUri.indexOf('@') + 10) +
      packageUri.slice(packageUri.indexOf('&path')),
  )
  return <StyledLink {...{ children, className, to }} />
}

const useListItemStyles = M.makeStyles((t) => ({
  root: {
    whiteSpace: 'nowrap',
    '&:hover $link': {
      display: 'inline',
    },
  },
  name: {
    flexGrow: 1,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  },
  link: {
    display: 'none',
    marginLeft: t.spacing(1),
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  },
  icon: {
    minWidth: t.spacing(5),
  },
  action: {
    right: 0, // https://github.com/mui/material-ui/issues/21722
  },
}))

interface ListItemProps {
  packageHandle?: PackageHandle
  bookmarks?: ReturnType<typeof Bookmarks.use>
  handle: Model.S3.S3ObjectLocation
  className: string
  onClear: () => void
}

function ListItem({
  bookmarks,
  className,
  handle,
  onClear,
  packageHandle,
}: ListItemProps) {
  const classes = useListItemStyles()
  const isDir = s3paths.isDir(handle.key)
  const name = isDir ? s3paths.ensureSlash(basename(handle.key)) : basename(handle.key)
  const isBookmarked = bookmarks?.isBookmarked('main', handle)
  const toggleBookmark = () => bookmarks?.toggle('main', handle)
  return (
    <M.ListItem className={cx(classes.root, className)} disableGutters>
      {bookmarks && (
        <M.ListItemIcon className={classes.icon}>
          <M.IconButton size="small" onClick={toggleBookmark}>
            <M.Icon fontSize="small">
              {isBookmarked ? 'turned_in' : 'turned_in_not'}
            </M.Icon>
          </M.IconButton>
        </M.ListItemIcon>
      )}
      <M.ListItemIcon className={classes.icon}>
        <M.Icon fontSize="small">{isDir ? 'folder_open' : 'insert_drive_file'}</M.Icon>
      </M.ListItemIcon>
      <span className={classes.name}>{name}</span>
      <FileLink className={classes.link} packageHandle={packageHandle} handle={handle} />
      <M.ListItemSecondaryAction className={classes.action}>
        <M.IconButton size="small" onClick={onClear}>
          <M.Icon fontSize="small">clear</M.Icon>
        </M.IconButton>
      </M.ListItemSecondaryAction>
    </M.ListItem>
  )
}

const useStyles = M.makeStyles((t) => ({
  root: {
    background: t.palette.background.paper,
  },
  buttons: {
    display: 'flex',
  },
  button: {
    '& + &': {
      marginLeft: t.spacing(1),
    },
  },
  divider: {
    marginTop: t.spacing(2),
  },
  list: {
    background: t.palette.background.paper,
    maxHeight: '50vh',
    overflowY: 'auto',
  },
  item: {
    '&:hover': {
      background: t.palette.action.hover,
    },
  },
  listSection: {
    background: 'inherit',
  },
  auxList: {
    background: 'inherit',
    padding: 0,
  },
}))

interface DashboardProps {
  onClose: () => void
  packageHandle?: PackageHandle
}

export default function Dashboard({ onClose, packageHandle }: DashboardProps) {
  const classes = useStyles()

  const slt = useSelection()
  const lists = React.useMemo(() => toHandlesMap(slt.selection), [slt.selection])
  const handleClear = React.useCallback(() => {
    slt.clear()
    onClose()
  }, [slt, onClose])

  const handleRemove = React.useCallback(
    (prefixUrl: string, index: number) => {
      const { isEmpty } = slt.remove(prefixUrl, index)
      if (isEmpty) {
        onClose()
      }
    },
    [onClose, slt],
  )

  const bookmarksCtx = Bookmarks.use()
  const bookmarks = React.useMemo(
    () => (!packageHandle ? bookmarksCtx : null),
    [packageHandle, bookmarksCtx],
  )
  const hasSomethingToBookmark = React.useMemo(
    () =>
      bookmarks &&
      Object.values(lists).some((hs) =>
        hs.some((h) => !bookmarks.isBookmarked('main', h)),
      ),
    [bookmarks, lists],
  )
  const handleBookmarks = React.useCallback(() => {
    if (!bookmarks) return
    const handles = Object.values(lists).reduce((memo, hs) => [...memo, ...hs], [])
    if (hasSomethingToBookmark) {
      bookmarks.append('main', handles)
    } else {
      bookmarks.remove('main', handles)
    }
  }, [hasSomethingToBookmark, lists, bookmarks])

  return (
    <div className={classes.root}>
      <div className={classes.buttons}>
        {bookmarks && (
          <M.Button
            className={classes.button}
            color="primary"
            disabled={slt.isEmpty}
            onClick={handleBookmarks}
            size="small"
            variant="outlined"
          >
            {hasSomethingToBookmark ? 'Add to bookmarks' : 'Remove from bookmarks'}
          </M.Button>
        )}
        {!!packageHandle && (
          <FileView.ZipDownloadForm
            suffix={`package/${packageHandle.bucket}/${packageHandle.name}/${packageHandle.hash}`}
            className={classes.button}
            files={toHandlesList(slt.selection).map(({ key }) => key)}
          >
            <Buttons.Iconized label="Download selected" icon="archive" type="submit" />
          </FileView.ZipDownloadForm>
        )}
        <M.Button
          className={classes.button}
          color="primary"
          disabled={slt.isEmpty}
          onClick={handleClear}
          size="small"
          variant="outlined"
        >
          Clear selection
        </M.Button>
      </div>
      <M.Divider className={classes.divider} />
      {!slt.isEmpty ? (
        <M.List dense disablePadding className={classes.list}>
          {Object.entries(lists).map(([prefixUrl, handles]) =>
            handles.length ? (
              <li className={classes.listSection} key={prefixUrl}>
                <ul className={classes.auxList}>
                  <M.ListSubheader disableGutters>{prefixUrl}</M.ListSubheader>
                  <M.List dense disablePadding>
                    {handles.map((handle, index) => (
                      <ListItem
                        key={handle.key}
                        bookmarks={bookmarks}
                        className={classes.item}
                        handle={handle}
                        onClear={() => handleRemove(prefixUrl, index)}
                        packageHandle={packageHandle}
                      />
                    ))}
                  </M.List>
                </ul>
              </li>
            ) : null,
          )}
        </M.List>
      ) : (
        <EmptyState />
      )}
    </div>
  )
}
