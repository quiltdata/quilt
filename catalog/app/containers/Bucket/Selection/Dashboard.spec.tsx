import * as React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

import Dashboard from './Dashboard'
import { Provider, useSelection } from './Provider'

type WithChildren = React.PropsWithChildren<{}>

// Mock Material-UI components
jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  Button: ({ children }: WithChildren) => <button role="button">{children}</button>,
  Divider: () => <hr />,
  List: ({ children }: WithChildren) => <ul>{children}</ul>,
  ListItem: ({ children }: WithChildren) => <li>{children}</li>,
  ListItemIcon: ({ children }: WithChildren) => <figure>{children}</figure>,
  ListItemSecondaryAction: ({ children }: WithChildren) => <div>{children}</div>,
  ListSubheader: ({ children }: WithChildren) => <li role="title">{children}</li>,
  Icon: ({ children }: WithChildren) => <figure>{children}</figure>,
  IconButton: ({ children }: WithChildren) => <button role="button">{children}</button>,
  Typography: ({ children }: WithChildren) => <p>{children}</p>,
}))

// Mock dependencies
jest.mock('../FileView', () => ({
  ZipDownloadForm: ({ children, ...props }: any) => (
    <div data-testid="zip-download-form" {...props}>
      {children}
    </div>
  ),
}))

jest.mock('utils/NamedRoutes', () => ({
  ...jest.requireActual('utils/NamedRoutes'),
  use: () => ({
    urls: {
      bucketDir: (bucket: string, key: string) => `/bucket/${bucket}/tree/${key}`,
      bucketFile: (bucket: string, key: string) => `/bucket/${bucket}/file/${key}`,
      bucketPackageTree: (bucket: string, name: string, hash: string, key: string) =>
        `/package/${bucket}/${name}/${hash}/tree/${key}`,
    },
  }),
}))

interface TestWrapperProps {
  children: React.ReactNode
  initialSelection?: any[]
}

// Test component that allows setting up initial selection state
function TestWrapper({ children, initialSelection }: TestWrapperProps) {
  return (
    <MemoryRouter>
      <Provider>
        <SelectionSetup initialSelection={initialSelection} />
        {children}
      </Provider>
    </MemoryRouter>
  )
}

function SelectionSetup({ initialSelection }: { initialSelection?: any[] }) {
  const selection = useSelection()
  const [hasSetup, setHasSetup] = React.useState(false)

  React.useEffect(() => {
    if (initialSelection && !hasSetup) {
      initialSelection.forEach(({ items, bucket, path, filter }) => {
        selection.merge(items, bucket, path, filter)
      })
      setHasSetup(true)
    }
  }, [initialSelection, selection, hasSetup])

  return null
}

describe('containers/Bucket/Selection/Dashboard', () => {
  it.skip('should render empty state when no selection', () => {
    const { container } = render(
      <TestWrapper>
        <Dashboard onClose={jest.fn()} />
      </TestWrapper>,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render with single file selection', () => {
    const initialSelection = [
      {
        items: [{ logicalKey: 'document.txt' }],
        bucket: 'test-bucket',
        path: 'files/',
      },
    ]

    const { container } = render(
      <TestWrapper initialSelection={initialSelection}>
        <Dashboard onClose={jest.fn()} />
      </TestWrapper>,
    )

    expect(container).toMatchSnapshot()
  })

  // it.skip('should render with multiple files in single directory', () => {
  //   const initialSelection = [
  //     {
  //       items: [
  //         { logicalKey: 'document1.txt' },
  //         { logicalKey: 'document2.pdf' },
  //         { logicalKey: 'subfolder/nested.json' },
  //       ],
  //       bucket: 'my-bucket',
  //       path: 'data/',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render with files from multiple directories', () => {
  //   const initialSelection = [
  //     {
  //       items: [{ logicalKey: 'file1.txt' }, { logicalKey: 'file2.txt' }],
  //       bucket: 'bucket-a',
  //       path: 'dir1/',
  //     },
  //     {
  //       items: [{ logicalKey: 'config.json' }],
  //       bucket: 'bucket-b',
  //       path: 'config/',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render with directory selection', () => {
  //   const initialSelection = [
  //     {
  //       items: [{ logicalKey: 'images/' }, { logicalKey: 'documents/' }],
  //       bucket: 'media-bucket',
  //       path: '',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render with package handle', () => {
  //   const packageHandle = {
  //     bucket: 'package-bucket',
  //     name: 'my-package',
  //     hash: 'abc123def456',
  //   }

  //   const initialSelection = [
  //     {
  //       items: [{ logicalKey: 'data.csv' }, { logicalKey: 'metadata.json' }],
  //       bucket: 'package-bucket',
  //       path: 'dataset/',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} packageHandle={packageHandle} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render with mixed file types and directories', () => {
  //   const initialSelection = [
  //     {
  //       items: [
  //         { logicalKey: 'README.md' },
  //         { logicalKey: 'src/' },
  //         { logicalKey: 'package.json' },
  //         { logicalKey: 'tests/' },
  //         { logicalKey: 'build/output.js' },
  //       ],
  //       bucket: 'project-bucket',
  //       path: '',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render with filtered selection', () => {
  //   const initialSelection = [
  //     {
  //       items: [{ logicalKey: 'logs/error.log' }, { logicalKey: 'logs/access.log' }],
  //       bucket: 'app-bucket',
  //       path: 'application/',
  //       filter: 'logs/',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })

  // it.skip('should render without bookmarks when package handle is present', () => {
  //   const packageHandle = {
  //     bucket: 'no-bookmark-bucket',
  //     name: 'test-package',
  //     hash: 'xyz789abc123',
  //   }

  //   const initialSelection = [
  //     {
  //       items: [{ logicalKey: 'data.txt' }],
  //       bucket: 'no-bookmark-bucket',
  //       path: '',
  //     },
  //   ]

  //   const { container } = render(
  //     <TestWrapper initialSelection={initialSelection}>
  //       <Dashboard onClose={mockOnClose} packageHandle={packageHandle} />
  //     </TestWrapper>,
  //   )

  //   expect(container).toMatchSnapshot()
  // })
})
