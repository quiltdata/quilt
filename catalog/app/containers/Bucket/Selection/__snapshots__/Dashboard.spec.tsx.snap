// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Selection/Dashboard should render empty state when no selection 1`] = `
<div>
  <div>
    <div>
      <button
        role="button"
      >
        Remove from bookmarks
      </button>
      <button
        role="button"
      >
        Clear selection
      </button>
    </div>
    <hr />
    <p>
      Nothing selected
    </p>
  </div>
</div>
`;

exports[`containers/Bucket/Selection/Dashboard should render with single file selection 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-buttons-2"
    >
      <button
        role="button"
      >
        Clear selection
      </button>
    </div>
    <hr />
    <ul>
      <li
        class="makeStyles-listSection-7"
      >
        <ul
          class="makeStyles-auxList-8"
        >
          <li
            role="title"
          >
            s3://test-bucket/files
          </li>
          <ul>
            <li>
              <figure>
                <figure>
                  insert_drive_file
                </figure>
              </figure>
              <span
                class="makeStyles-name-11"
              >
                document.txt
              </span>
              <a
                class="makeStyles-link-12 makeStyles-root-15"
                href="/bucket/test-bucket/file/files/document.txt"
              >
                s3://test-bucket/files/document.txt
              </a>
              <div>
                <button
                  role="button"
                >
                  <figure>
                    clear
                  </figure>
                </button>
              </div>
            </li>
          </ul>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;
